crypto/src/lib.rs

```diff
@@ -376,79 +501,132 @@ impl PublicKey {
 }
 
 // TODO: Enable in ffi_import
-#[cfg(feature = "std")]
 #[cfg(not(feature = "ffi_import"))]
 impl From<PrivateKey> for PublicKey {
     fn from(private_key: PrivateKey) -> Self {
-        Self::try_from_private(private_key).expect("can't fail for valid `PrivateKey`")
+        let digest_function = private_key.algorithm();
+        let key_gen_option = Some(KeyGenOption::FromPrivateKey(private_key));
+
+        match digest_function {
+            Algorithm::Ed25519 => {
+                PublicKey::Ed25519(ed25519::Ed25519Sha512::keypair(key_gen_option).0)
+            }
+            Algorithm::Secp256k1 => {
+                PublicKey::Secp256k1(secp256k1::EcdsaSecp256k1Sha256::keypair(key_gen_option).0)
+            }
+            Algorithm::BlsNormal => PublicKey::BlsNormal(bls::BlsNormal::keypair(key_gen_option).0),
+            Algorithm::BlsSmall => {
+                PublicKey::BlsSmall(Box::new(bls::BlsSmall::keypair(key_gen_option).0))
+            }
+        }
     }
 }
 
 ffi::ffi_item! {
     /// Private Key used in signatures.
-    #[derive(Clone, PartialEq, Eq, CopyGetters)]
-    #[cfg_attr(not(feature="ffi_import"), derive(DebugCustom, Display, Serialize))]
-    #[cfg_attr(not(feature="ffi_import"), debug(fmt = "{{digest: {digest_function}, payload: {}}}", "hex::encode_upper(payload)"))]
-    #[cfg_attr(not(feature="ffi_import"), display(fmt = "{}", "hex::encode_upper(payload)"))]
-    pub struct PrivateKey {
-        /// Digest function
-        #[getset(get_copy = "pub")]
-        digest_function: Algorithm,
-        /// Key payload
-        #[serde(with = "hex::serde")]
-        payload: ConstVec<u8>,
+    #[derive(Clone)]
+    #[cfg_attr(all(feature = "ffi_export", not(feature = "ffi_import")), ffi_type(opaque))]
+    #[allow(missing_docs)]
+    pub enum PrivateKey {
+        Ed25519(Box<ed25519::PrivateKey>),
```

Comment:
I don't quite like seeing a `Box` here

Reply:
`ed25519::PrivateKey` contains a lot of math data, it weights `224` bytes, while with this `Box` our main `PrivataKey` weights just `40`.

Not sure what else we can do there

Reply:
that's a big difference. I see what you did now, what you could have done is to make a newtype wrapper with Box inside

Reply:
Yeap, just did it while trying to improve performance